test/functional/readpreference.test.js

```diff
@@ -587,4 +588,144 @@ describe('ReadPreference', function () {
       })
     });
   });
+
+  context('should enforce fixed primary read preference', function () {
+    const collectionName = 'ddl_collection';
+
+    beforeEach(function () {
+      const configuration = this.configuration;
+      const client = this.configuration.newClient(configuration.writeConcernMax(), {
+        useUnifiedTopology: true,
+        readPreference: 'primaryPreferred'
+      });
+      return withClient(client, (client, done) => {
+        const db = client.db(configuration.db);
+        db.addUser('default', 'pass', { roles: 'readWrite' }, () => {
+          db.createCollection('before_collection', () => {
+            db.createIndex(collectionName, { aloha: 1 }, done);
+          });
+        });
+      });
+    });
+
+    const methods = {
+      'Collection#createIndex': [{ quote: 'text' }],
+      'Db#createIndex': [collectionName, { quote: 'text' }],
+      'Db#addUser': ['thomas', 'pass', { roles: 'readWrite' }],
+      'Db#removeUser': ['default'],
+      'Db#createCollection': ['created_collection'],
+      'Db#dropCollection': ['before_collection'],
+      'Collection#dropIndex': ['aloha_1'],
+      'Collection#rename': ['new_name'],
+      'Db#dropDatabase': []
+    };
+
+    Object.keys(methods).forEach(operation => {
+      it(`${operation}`, {
+        metadata: {
+          requires: { topology: ['replicaset', 'sharded'] }
+        },
+        test: function () {
+          const configuration = this.configuration;
+          const client = this.configuration.newClient(configuration.writeConcernMax(), {
+            useUnifiedTopology: true,
+            readPreference: 'primaryPreferred'
+          });
+          return withClient(client, (client, done) => {
+            const db = client.db(configuration.db);
+            const args = methods[operation];
+            const [parentId, method] = operation.split('#');
+            const collection = db.collection(collectionName);
+            const parent = parentId === 'Collection' ? collection : parentId === 'Db' ? db : null;
+            const selectServerSpy = this.sinon.spy(Topology.prototype, 'selectServer');
+            const callback = err => {
+              expect(err).to.not.exist;
+              expect(selectServerSpy.called).to.equal(true);
+              if (typeof selectServerSpy.args[0][0] === 'function') {
+                expect(selectServerSpy)
+                  .nested.property('args[0][1].readPreference.mode')
+                  .to.equal(ReadPreference.PRIMARY);
+              } else {
+                expect(selectServerSpy)
+                  .nested.property('args[0][0].readPreference.mode')
+                  .to.equal(ReadPreference.PRIMARY);
+              }
+              done();
+            };
+            parent[method].apply(parent, [...args, callback]);
+          });
+        }
+      });
+    });
+  });
+
+  context('should enforce fixed primary read preference', function () {
```

Comment:
it looks like you have two copies of the same test additions

Reply:
doh, I was wrestling with this file / linting and manually doubled up.